### PR TITLE
Move read flow statistics to cursor

### PR DESCRIPTION
**Pull request by breeswish**
## What have you changed? (mandatory)

Previously our flow statistics is not very accurate. For example, when doing `near_seek`, the read during each `next()` attempt is not counted in the statistics. According to PD developers, these read flows should be counted however. Also, previously we self-implemented some read statistics in raw_scan and raw_batch_scan.

This PR moves the read flow statistics logic from mvcc layer into `Cursor` layer, the same as CFStatistics. In detail, when `key()` or `value()` is called, its read flow is counted. Subsequent calls will not be counted unless the cursor is moved (e.g. by `seek()`, `next()`, etc).

This change is extracted from the [Mvcc Refinement PR](https://github.com/pingcap/tikv/pull/3344).

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

CI

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

## Benchmark result if necessary (optional)

There is no notable performance difference after applying this PR's change:

master:

```
test storage::scan::bench_seek_ts                                    ... bench:      85,256 ns/iter (+/- 39,536)
test storage::scan::bench_seek_ts                                    ... bench:      82,645 ns/iter (+/- 13,326)
test storage::scan::bench_seek_ts                                    ... bench:      85,052 ns/iter (+/- 26,659)
test storage::scan::bench_seek_ts                                    ... bench:      87,078 ns/iter (+/- 10,894)
```

this PR:

```
test storage::scan::bench_seek_ts                                    ... bench:      84,566 ns/iter (+/- 17,616)
test storage::scan::bench_seek_ts                                    ... bench:      82,349 ns/iter (+/- 14,759)
test storage::scan::bench_seek_ts                                    ... bench:      82,558 ns/iter (+/- 14,742)
test storage::scan::bench_seek_ts                                    ... bench:      88,204 ns/iter (+/- 11,680)
test storage::scan::bench_seek_ts                                    ... bench:      82,717 ns/iter (+/- 12,728)
```

### CODE REVIEW

**Comment by breeswish** [comment]
@AndreMouche @zhangjinpeng1987 @MyonKeminta PTAL

File: src/storage/engine/mod.rs [hunk 1]
```diff
@@ -510,50 +525,83 @@ impl<I: Iterator> Cursor<I> {
...
+        statistics.seek_for_prev += 1;
+        self.is_key_read = false;
+        self.is_value_read = false;
+        self.iter.seek_for_prev(key)
+    }
+
     #[inline]
     pub fn next(&mut self, statistics: &mut CFStatistics) -> bool {
         statistics.next += 1;
+        self.is_key_read = false;
```
Comment (zhangjinpeng1987) [inline src/storage/engine/mod.rs]: It's better to extract a function `mark_unread`.

File: src/storage/engine/mod.rs [hunk 2]
```diff
@@ -319,6 +331,9 @@ pub struct Cursor<I: Iterator> {
     // the data cursor can be seen will be
     min_key: Option<Vec<u8>>,
     max_key: Option<Vec<u8>>,
+
+    is_key_read: bool,
```
Comment (zhangjinpeng1987) [inline src/storage/engine/mod.rs]: cur_key_has_read

File: src/raftstore/store/region_snapshot.rs [hunk 3]
```diff
@@ -539,13 +539,19 @@ mod tests {
             iter.reverse_seek(&Key::from_encoded(b"a7".to_vec()), &mut statistics)
                 .unwrap()
         );
-        let mut pair = (iter.key().to_vec(), iter.value().to_vec());
+        let mut pair = (
+            iter.key(&mut statistics).to_vec(),
```
Comment (siddontang) [inline src/raftstore/store/region_snapshot.rs]: em, can you extract them to a function?

**Comment by siddontang** [comment]
can you refer the statistics in the Cursor directly, it is weird to pass stats to key, value function every where.

**Comment by breeswish** [comment]
@siddontang We have already have the statistics *almost* everywhere in Cursor, for example, `seek()`, `seek_for_prev()`, `next()`, `prev()`, etc. 🤣 This PR just makes it no survivors. I have an idea for Cursor Statistics refinement after Mvcc refinement, by using Thread Local Storages to avoid passing variables everywhere, which should greatly help ease this situation. I will propose a standalone RFC for that change. However for this PR, let's accept it for now, as a patch solution before total refinement.

File: src/storage/engine/mod.rs [hunk 4]
```diff
@@ -328,9 +343,18 @@ impl<I: Iterator> Cursor<I> {
             scan_mode: mode,
             min_key: None,
             max_key: None,
+
+            cur_key_has_read: false,
+            cur_value_has_read: false,
         }
     }
 
+    /// Mark key and value as unread. It will be invoked once cursor is moved.
```
Comment (zhangjinpeng1987) [inline src/storage/engine/mod.rs]: Please inline this function.

Comment (breeswish) [reply src/storage/engine/mod.rs]: done

File: src/raftstore/store/region_snapshot.rs (hunk 3 above)
Comment (breeswish) [reply src/raftstore/store/region_snapshot.rs]: don't worry. many of them are removed in Mvcc refinement PR later.

**Pull request merged**